### PR TITLE
Support meta in apply gufunc

### LINKED ISSUE #6358: support meta in apply_gufunc
**What happened**:

`dask.array.apply_ufunc` does not accept a `meta` kwarg.

**Motivation**

There is a PR (https://github.com/pydata/xarray/pull/4060) updating `xarray.apply_ufunc` to use `dask.array.apply_gufunc`. Currently `apply_ufunc` used `blockwise` which accepted a `meta` kwarg but now we see that `apply_gufunc` does not accept one.

I don't have an example that fails with the current code. But it seems like a good idea to let the user specify `meta` in case the automated inference in the following lines doesn't work
https://github.com/dask/dask/blob/ad5006fc70a256ceb389a3e42b6f303c61c1721d/dask/array/gufunc.py#L420-L440

### PR DESCRIPTION
**Pull request by joshreback**
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

closes #6358

### CODE REVIEW

File: dask/array/tests/test_gufunc.py [hunk 1]
```diff
@@ -615,3 +616,18 @@ def stats(x):
...
+
+    a = da.random.normal(size=(10, 20, 30), chunks=(5, 5, 30))
+    meta = (np.ones(0, dtype=np.float64), np.ones(0, dtype=np.float32))
+    result = apply_gufunc(stats, "(i)->(),()", a, meta=meta)
+    mean, std = result
+    assert isinstance(result, tuple)
+    assert mean.compute().shape == (10, 20)
+    assert std.compute().shape == (10, 20)
+    for expected, actual in zip(meta, result):
+        assert expected.dtype == actual._meta.dtype
```
Comment (mrocklin) [inline dask/array/tests/test_gufunc.py:633]: Maybe we could do something like 

```python
expected = stats(a.compute())
assert_eq(expected[0], result[0])
assert_eq(expected[1], result[1])
```

Using assert_eq checks lots of things, like shape, dtype, and other metadata.  It's generally good testing hygiene in Dask to use it when possible.

**Comment by mrocklin** [comment]
This seems good to me.  Thank you for adding this @joshreback .  I'm good to merge as-is, although if you wanted to implement the suggestion around assert_eq I think that that would be marginally better.

File: dask/array/tests/test_gufunc.py (hunk 1 above)
Comment (joshreback) [reply dask/array/tests/test_gufunc.py]: Thanks for the feedback! I switched to using `assert_eq` as suggested, which actually exposed the fact that passing `np.float32` as the dtype for the second meta was not actually surviving after calling `compute`. I had to pass in the dtype in the gufunc and in the meta for it to actually take effect. The test illustrates that now, but let me know if you think it makes sense to add more tests to illustrate the usage.

**Comment by joshreback** [comment]
> This seems good to me. Thank you for adding this @joshreback . I'm good to merge as-is, although if you wanted to implement the suggestion around assert_eq I think that that would be marginally better.

Hi @mrocklin - are there any other feedback or suggestions that I should address here?

**Comment by TomAugspurger** [comment]
I think we're good, thanks for following up.

**Comment by mrocklin** [comment]
Thanks!

**Comment by jakirkham** [comment]
Very cool! Thanks Josh (and everyone who reviewed)! 😄

cc @pentschev (for vis)

**Comment by pentschev** [comment]
It's really awesome to see more `_meta` support, thanks @joshreback !

**Pull request merged**